[ADR -44] Lite Client with Weak Subjectivity (#3795)
tendermint/tendermint: docs/architecture/adr-044-lite-client-with-weak-subjectivity.md

# ADR 044: Lite Client with Weak Subjectivity

## Changelog
* 13-07-2019: Initial draft
* 14-08-2019: Address cwgoes comments

## Context

The concept of light clients was introduced in the Bitcoin white paper. It
describes a watcher of distributed consensus process that only validates the
consensus algorithm and not the state machine transactions within.

Tendermint light clients allow light weight devices and other blockchains to
efficiently verify the consensus of a Tendermint blockchain. This forms the
basis of safe and efficient state synchronization for new network nodes and
inter-blockchain communication (where a light client of one Tendermint instance
runs in another chain's state machine).

In a network that is expected to reliably punish validators for misbehavior
through punishments and where the validator set changes
infrequently, clients can take advantage of this assumption to safely
synchronize a lite client without downloading the intervening headers.

Light clients (and full nodes) operating in the Proof Of Stake context need a
trusted block height from a trusted source that is no older than 1 unbonding

> window. This is called “weak subjectivity”.

Reviewer: ```suggestion
window plus a configurable evidence submission synchrony bound. This is called “weak subjectivity”.
```

> Weak Subjectivity is required in Proof of Stake blockchains because it is

Reviewer: ```suggestion
Weak subjectivity is required in Proof of Stake blockchains because it is
```

costless for an attacker to buy up voting keys that are no longer bonded and
fork the network at some point in its prior history. See Vitalik’s post at
[Proof of Stake: How I Learned to Love Weak
Subjectivity](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/).

Currently, Tendermint provides a lite client implementation in the
[lite](https://github.com/tendermint/tendermint/tree/master/lite) package. This

> lite client implements a bisecting algorithm that tries to use a binary search

Reviewer: ```suggestion
lite client implements a bisection algorithm that tries to use a binary search
```

to find the minimum number of block headers where the validator set voting

> power changes are less than < 1/3rd. This interface does not support weak

Reviewer: This sentence is confusing, do you mean that this interface does not support starting with a recent trusted header?

subjectivity at this time. The Cosmos SDK also does not support counterfactual
slashing, nor does the lite client have any capacity to report evidence making
these systems *theoretically unsafe*.

NOTE: Tendermint provides a somewhat different (stronger) light client model
than Bitcoin under eclipse, since the eclipsing node(s) can only fool the light
client if they have two-thirds of the private keys from the last root-of-trust.

## Decision

### The Weak Subjectivity Interface

> Add the weak subjectivity interface for when a new light client connect to the

Reviewer: ```suggestion
Add the weak subjectivity interface for when a new light client connects to the
```

network or when a light client that has been offline for longer than the
unbonding period connects to the network. Specifically, the node needs to
initialize the following structure before syncing from user input:

```
type TrustOptions struct {
    // Required: only trust commits up to this old.
    // Should be equal to the unbonding period minus some delta for evidence reporting.
    TrustPeriod time.Duration `json:"trust-period"`

    // Option 1: TrustHeight and TrustHash can both be provided
    // to force the trusting of a particular height and hash.
    // If the latest trusted height/hash is more recent, then this option is
    // ignored.
    TrustHeight int64  `json:"trust-height"`
    TrustHash   []byte `json:"trust-hash"`

    // Option 2: Callback can be set to implement a confirmation
    // step if the trust store is uninitialized, or expired.
    Callback func(height int64, hash []byte) error
}
```

The expectation is the user will get this information from a trusted source

> like a validator, https://cosmos.network or other. A more user friendly

Reviewer: ```suggestion
like a validator, a friend, or a secure website. A more user friendly
```

> solution with trust tradeoff is that we establish an https based protocol with

Reviewer: ```suggestion
solution with trust tradeoffs is that we establish an https based protocol with
```

a default end point that populates this information. Also an on-chain registry
of roots-of-trust (e.g. on the Cosmos Hub) seems likely in the future.

### Linear Verification

> The linear verification of the light client requires downloading all headers

Reviewer: ```suggestion
The linear verification algorithm requires downloading all headers
```

between the `TrustHeight` and the `LatestHeight`. The lite client downloads the
full header for the provided `TrustHeight` and then proceeds to download `N+1`
headers and applies the [Tendermint validation
rules](https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/blockchain.md#validation)
to each block.

### Bisecting Verification

Bisecting Verification is a more bandwidth and compute intensive mechanism that
in the most optimistic case requires a light client to only download two block
headers to come into synchronization.

> The Bisection algorithm proceeds in the following fashion. The client downloads

Reviewer: ```suggestion
The bisection algorithm proceeds in the following fashion. The client downloads
```

and verifies the full block header for `TrustHeight` and then  fetches

> `LastestHeight` blocker header. The client then verifies the `LatestHeight`

Reviewer: ```suggestion
`LatestHeight` blocker header. The client then verifies the `LatestHeight`
```

header. Finally the client attempts to verify the `LatestHeight` header with

> voting powers taken from `NextValdiatorSet` in the `TrustHeight` header. This

Reviewer: ```suggestion
voting powers taken from `NextValidatorSet` in the `TrustHeight` header. This
```

> verification will succeed if the validators from `TrustHeight` still have > 2/3

Reviewer: This is a different threshold than mentioned above. I think we can vary this - see discussion in Josef's PR (https://github.com/tendermint/tendermint/pull/3841). Certainly we should reference that document.

Reviewer 2: Linked. Also asked a question about `trustlevel` https://github.com/tendermint/tendermint/pull/3841/files#r314187482

+1 of voting power in the `LatestHeight`. If this succeeds, the client is fully
synchronized. If this fails, then following Bisection Algorithm should be
executed.

The Client tries to download the block at the mid-point block between
`LatestHeight` and `TrustHeight` and attempts that same algorithm as above
using `MidPointHeight` instead of `LatestHeight`. In the case the of failure,
recursively perform the `MidPoint` verification until success then start over
with an updated `NextValidatorSet` and `TrustHeight`.

> An optimal client will cache headers from previous iterations of recursion to

Reviewer: Why is this necessary? Are there cases in the search when we request the same header again?

Reviewer 2: Maybe in cases where the light client detects malicious behavior during bisection and switches to other full-node?

Reviewer: But then we wouldn't want to cache the headers...

Reviewer 2: Right...

Reviewer 2: @cwgoes `^^^

minimize network usage.

## Status

Accepted.

## Consequences

### Positive

* light client which is safe to use (it can go offline, but not for too long)

### Negative

> * social consensus can be prone to errors (for cases where a new light client

Reviewer: A cost is the complexity of bisection. The social consensus cost of weak subjectivity is unavoidable.

# Tendermint Lite Client with Weak Subjectivity ADR

> 

Reviewer 3: ```suggestion
Changelog:
- 13-07-2019: Initial Draft
```

# Tendermint Lite Client with Weak Subjectivity ADR

Changelog:
- 13-07-2019: Initial Draft
## Context

> The concept of light clients was introduced in the Bitcoin white paper. It describes a watcher of distributed consensus process that only validates the consensus algorithm and not the state machine transactions within.

Reviewer: Two things to note:
- Light clients are expected to agree with full nodes on the canonical chain
- Tendermint provides a somewhat different (stronger) light client model under eclipse, since the eclipsing node(s) can only fool the light client if they have two-thirds of the private keys from the last root of trust.

> Tendermint light clients allow light weight devices and other blockchains to efficiently verify the consensus of a Tendermint blockchain. This forms the basic of safe and efficient state synchronization for new network nodes and InterBlockchain Communication.

Reviewer: ```suggestion
Tendermint light clients allow light weight devices and other blockchains to efficiently verify the consensus of a Tendermint blockchain. This forms the basis of safe and efficient state synchronization for new network nodes and inter-blockchain communication (where a light client of one Tendermint instance runs in another chain's state machine).
```

> In a network that is expected to reliably punish validators for misbehavior through punishments or where the validator set is largely trusted and changes infrequently, clients can take advantage of this assumption to safely synchronize a lite client without downloading the intervening headers.

Reviewer: "largely trusted" is a bit vague. I think both conditions are necessary - only if misbehaviour is punished and validator set changes are infrequent can a light client skip most of the headers. Note also that the state machine can enforce limits on validator set liquidity, which might be useful to provide tighter efficiency/safety bounds.

> Light clients (and full nodes) operating in the Proof Of Stake context need a trusted block height from a trusted source that is no older than 1 unbending window. This is called “weak subjectivity”

Reviewer 2: `unbending` -> `unbonding`

> Light clients (and full nodes) operating in the Proof Of Stake context need a trusted block height from a trusted source that is no older than 1 unbending window. This is called “weak subjectivity”

Reviewer: We should explain why one unbonding window (and I think it should be one unbonding window minus delta, where delta is a parameter set according to the expected synchrony bounds for the light client to be able to report evidence)

Currently, Tendermint provides a lite client implementation in the [tendermint/lite](https://github.com/tendermint/tendermint/tree/master/lite) directory. This lite client implements a bisecting algorithm that tries to use a binary search to find the minimum number of block headers where the validator set voting power changes are less than <1/3rd. This interface does not support weak subjectivity at this time. The Cosmos SDK also does not support counterfactual slashing or does the lite client have any capacity to report evidence making these systems theoretically unsafe.

## Decision 

### The Weak Subjectivity Interface.

> Weak Subjectivity is required in Proof of Stake blockchains because it is costless for an attack to buy up voting keys that are no longer bonded and fork the network at some point it’s history. See Vitalik’s post at [Proof of Stake: How I Learned to Love Weak Subjectivity](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/)

Reviewer: ```suggestion
Weak Subjectivity is required in Proof of Stake blockchains because it is costless for an attacker to buy up voting keys that are no longer bonded and fork the network at some point in its prior history. See Vitalik’s post at [Proof of Stake: How I Learned to Love Weak Subjectivity](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity/)
```

> This requires providing a user interface for when new nodes connect to the network or when a node that has been offline for longer than the unbending period connects to the network. 

Reviewer 2: `unbending` -> `unbonding`

Reviewer: We should totally call it the `unbending` period though.

Reviewer 2: hm.. okay. haven't seen unbending word before

Reviewer: (just kidding)

> Specifically, the node node needs to initialize the following structure before synching from user input. The expectation is the user will get this information from a trusted source like a validator, https://cosmos.network or other.  A more user friendly solution with trust tradeoff is that we establish an https based protocol with a default end point that populates this information. 

Reviewer 2: `node node` -> `node`

> Specifically, the node node needs to initialize the following structure before synching from user input. The expectation is the user will get this information from a trusted source like a validator, https://cosmos.network or other.  A more user friendly solution with trust tradeoff is that we establish an https based protocol with a default end point that populates this information. 

Reviewer: I would note that an on-chain registry of roots-of-trust (e.g. on the Cosmos Hub) seems likely in the future.

```
type TrustOptions struct {

> // Required: only trust commits up to this old.

Reviewer 2: formatting is off

> // Should be equal to the unbonding period.

Reviewer: ```suggestion
// Should be equal to the unbonding period minus some delta for evidence reporting.
```

TrustPeriod time.Duration `json:"trust-period"`

// Option 1: TrustHeight and TrustHash can both be provided
// to force the trusting of a particular height and hash.
// If the latest trusted height/hash is more recent, then this option is
// ignored.
TrustHeight int64  `json:"trust-height"`
TrustHash   []byte `json:"trust-hash"`

// Option 2: Callback can be set to implement a confirmation
// step if the trust store is uninitialized, or expired.
Callback func(height int64, hash []byte) error
}
```

## Linear Verification

> The linear verification of the light client requires downloading all headers between the `TrustHeight` and the `LatestHeight`. The lite client downloads the full header for the provided `TrustHeight` and then proceeds to download `N+1` headers and applies the [Tendermint validation rules](https://github.com/tendermint/tendermint/blob/master/docs/spec/blockchain/blockchain.md#validation) to each block. 

Reviewer: Should we combine https://github.com/tendermint/tendermint/pull/3796/files here?

## Bisecting Verification

> Bisecting Verification is a more bandwidth and compute intensive mechanism that in the most optimistic case requires a light client to only download two block headers to come into synchronization. 

Reviewer: It's a less bandwidth and compute-intensive mechanism, right?

We can bound the number of headers precisely if we make assumptions about validator set liquidity.

> The Bisection algorithm proceeds in the following fashion. The client downloads  and verifies the full block header for `TrustHeight` and then  fetches  `LastestHeight` blocker header.  The client then verifies the `LatestHeight` header.  Finally the client attempts to verify the `LatestHeight` header with voting powers taken from `NextValdiatorSet` in the `TrustHeight` header. This will verification will succeed if that validators from `TrustHeight` still have > 2/3 +1 of voting power in the `LatestHeight`. If this succeeds, the client is fully synchronized. If this fails, then following Bisection Algorithm should be followed.

Reviewer 4: ```suggestion
The Bisection algorithm proceeds in the following fashion. The client downloads  and verifies the full block header for `TrustHeight` and then  fetches  `LastestHeight` blocker header.  The client then verifies the `LatestHeight` header.  Finally the client attempts to verify the `LatestHeight` header with voting powers taken from `NextValdiatorSet` in the `TrustHeight` header. This verification will succeed if the validators from `TrustHeight` still have > 2/3 +1 of voting power in the `LatestHeight`. If this succeeds, the client is fully synchronized. If this fails, then following Bisection Algorithm should be executed.
```

> The Bisection algorithm proceeds in the following fashion. The client downloads  and verifies the full block header for `TrustHeight` and then  fetches  `LastestHeight` blocker header.  The client then verifies the `LatestHeight` header.  Finally the client attempts to verify the `LatestHeight` header with voting powers taken from `NextValdiatorSet` in the `TrustHeight` header. This will verification will succeed if that validators from `TrustHeight` still have > 2/3 +1 of voting power in the `LatestHeight`. If this succeeds, the client is fully synchronized. If this fails, then following Bisection Algorithm should be followed.

Reviewer 5: Note that +2/3 in the new validator set is not sufficient to establish trust of the new validator set as total voting power of new validators and potential adversary set of the old set could be bigger than +1/3 of total voting power. See https://github.com/tendermint/tendermint/pull/3710.

Author: So it seems like we should say that in order for bisection to take place, the total voting power of the new block must be greater or equal to the total voting power of the last trusted block to prevent this attack correct?

Reviewer 5: I think it's more complex than that. Total voting power of new validators together with 1/3 of voting power of old validators that are also present in new validator set must be in total at most 1/3 of voting power in the new val set.

Author: Okay here is my proposed check.

After current checks for if `LatestHeight` is valid in it's own terms, if `LatestHeight` is valid in terms of `TrustedHeight.NextValidatorSet` we compute the following. Take the sum of the following, for each validator `abs((LatestHeightVotingPower/LatestHeightTotalPower)-TrustedHeightVotingPower/TrustedHeightTotalPower))`. if this value is greater than 1/3rd, fail for the height.

Author: I'm still somewhat split if we want to have this check in the presence of counterfactual slashing.

It does confuse the incentive layers a bit, but counterfactual slashing does ensure that any attack using this kind of change of voting powers is still slashable.

Reviewer 5: I still have trouble to see link between counterfactual slashing and the checks about validator power change, i.e., how we can get rid of checks in the presence of counterfactual slashing.

Author: So an attacker who wants to exploit this vulnerability needs signatures from 2/3rd from the TrustedSet. in attack these signatures are either equivocations or counterfactual signatures.

> The Client tries to download the block at the mid-point block between `LatestHeight` and `TrustHeight` and attempts that same algorithm as above using `MidPointHeight` instead of `LatestHeight`. In the case the of failure, recursively perform the `MidPoint` verification until success then start over with an updated `NextValidatorSet` and `TrustHeight`.

Reviewer: I would note that it's safe to ask the full node to provide the specific headers that the client needs, given the root of trust height that it has. Should we do that instead? The full node has to track a bit more, but further reduces bandwidth / compute for the light client.
